File: ios/Sources/NoiseProtocol/Benchmark.swift
import Foundation

public struct BenchmarkResult: Sendable {
    public let name: String
    public let iterations: Int
    public let totalNs: Int64
    public let opsPerSec: Double
    public let avgNs: Double
}

public struct BenchmarkComparison: Sendable {
    public let name: String
    public let baselineOpsPerSec: Double
    public let currentOpsPerSec: Double
    public let speedup: Double
    public let improved: Bool
}

public struct ComparisonReport: Sendable {
    public let comparisons: [BenchmarkComparison]

    public func toSummary() -> String {
        var lines = [String]()
        let header = "Benchmark" + String(repeating: " ", count: 21) +
                     "       Baseline        Current    Speedup   Status"
        lines.append(header)
        lines.append(String(repeating: "-", count: 83))
        for c in comparisons {
            let status: String
            if c.speedup > 1.05 { status = "✓ FASTER" }
            else if c.speedup < 0.95 { status = "✗ SLOWER" }
            else { status = "= SAME" }
            let namePadded = c.name.padding(toLength: 30, withPad: " ", startingAt: 0)
            let line = "\(namePadded) \(String(format: "%13.1f %13.1f %9.2fx", c.baselineOpsPerSec, c.currentOpsPerSec, c.speedup)) \(status)"
            lines.append(line)
        }
        return lines.joined(separator: "\n")
    }
}

public struct BenchmarkSuite: Sendable {
    public let platform: String
    public let timestamp: String
    public let results: [BenchmarkResult]

    public func toJson() -> String {
        var lines = [String]()
        lines.append("{")
        lines.append("  \"platform\": \"\(platform)\",")
        lines.append("  \"timestamp\": \"\(timestamp)\",")
        lines.append("  \"results\": [")
        for (i, r) in results.enumerated() {
            lines.append("    {")
            lines.append("      \"name\": \"\(r.name)\",")
            lines.append("      \"iterations\": \(r.iterations),")
            lines.append("      \"totalNs\": \(r.totalNs),")
            lines.append("      \"opsPerSec\": \(r.opsPerSec),")
            lines.append("      \"avgNs\": \(r.avgNs)")
            lines.append("    }" + (i < results.count - 1 ? "," : ""))
        }
        lines.append("  ]")
        lines.append("}")
        return lines.joined(separator: "\n")
    }

    public func saveToFile(_ url: URL) throws {
        try toJson().write(to: url, atomically: true, encoding: .utf8)
    }

    public static func loadFromFile(_ url: URL) throws -> BenchmarkSuite {
        let json = try String(contentsOf: url, encoding: .utf8)
        return fromJson(json)
    }

    public static func fromJson(_ json: String) -> BenchmarkSuite {
        let platform = json.extractString("platform")
        let timestamp = json.extractString("timestamp")

        var results = [BenchmarkResult]()
        guard let resultsStart = json.range(of: "\"results\":") else { return BenchmarkSuite(platform: platform, timestamp: timestamp, results: []) }
        let afterResults = String(json[resultsStart.upperBound...])
        guard let arrayStart = afterResults.firstIndex(of: "["),
              let arrayEnd = afterResults.lastIndex(of: "]") else { return BenchmarkSuite(platform: platform, timestamp: timestamp, results: []) }

        let arrayContent = String(afterResults[afterResults.index(after: arrayStart)..<arrayEnd])

        // Split on }...{ pattern (with optional whitespace/newlines/commas between)
        let objectStrings = arrayContent.components(separatedBy: "}")
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines).replacingOccurrences(of: ",", with: "").replacingOccurrences(of: "{", with: "") }
            .filter { !$0.isEmpty }

        for obj in objectStrings {
            results.append(BenchmarkResult(
                name: obj.extractString("name"),
                iterations: obj.extractInt("iterations"),
                totalNs: obj.extractInt64("totalNs"),
                opsPerSec: obj.extractDouble("opsPerSec"),
                avgNs: obj.extractDouble("avgNs")
            ))
        }

        return BenchmarkSuite(platform: platform, timestamp: timestamp, results: results)
    }

    public func compareToBaseline(_ baseline: BenchmarkSuite) -> ComparisonReport {
        var baselineByName = [String: BenchmarkResult]()
        for r in baseline.results { baselineByName[r.name] = r }

        let comparisons = results.map { current -> BenchmarkComparison in
            let baseOps = baselineByName[current.name]?.opsPerSec ?? current.opsPerSec
            let speedup = baseOps > 0 ? current.opsPerSec / baseOps : 1.0
            return BenchmarkComparison(
                name: current.name,
                baselineOpsPerSec: baseOps,
                currentOpsPerSec: current.opsPerSec,
                speedup: speedup,
                improved: speedup > 1.0
            )
        }
        return ComparisonReport(comparisons: comparisons)
    }
}

public final class BenchmarkRunner: Sendable {
    public let warmupIterations: Int
    public let measureIterations: Int

    public init(warmupIterations: Int = 100, measureIterations: Int = 1000) {
        self.warmupIterations = warmupIterations
        self.measureIterations = measureIterations
    }

    public func run(_ name: String, block: () throws -> Void) rethrows -> BenchmarkResult {
        // Warmup
        for _ in 0..<warmupIterations { try block() }

        // Measure
        let start = DispatchTime.now()
        for _ in 0..<measureIterations { try block() }
        let end = DispatchTime.now()

        let totalNs = Int64(end.uptimeNanoseconds - start.uptimeNanoseconds)
        let avgNs = Double(totalNs) / Double(measureIterations)
        let opsPerSec = totalNs > 0 ? Double(measureIterations) / (Double(totalNs) / 1_000_000_000.0) : 0.0

        return BenchmarkResult(
            name: name,
            iterations: measureIterations,
            totalNs: totalNs,
            opsPerSec: opsPerSec,
            avgNs: avgNs
        )
    }
}

// MARK: - JSON Parsing Helpers

private extension String {
    func extractString(_ key: String) -> String {
        guard let range = self.range(of: "\"\(key)\"\\s*:\\s*\"([^\"]*?)\"", options: .regularExpression) else { return "" }
        let match = String(self[range])
        guard let valueStart = match.range(of: ": \"") else { return "" }
        let value = String(match[valueStart.upperBound...]).dropLast()
        return String(value)
    }

    func extractInt(_ key: String) -> Int {
        guard let range = self.range(of: "\"\(key)\"\\s*:\\s*(\\d+)", options: .regularExpression) else { return 0 }
        let match = String(self[range])
        guard let colonRange = match.range(of: ": ") else { return 0 }
        let numStr = match[colonRange.upperBound...].trimmingCharacters(in: .whitespaces)
        return Int(numStr) ?? 0
    }

    func extractInt64(_ key: String) -> Int64 {
        guard let range = self.range(of: "\"\(key)\"\\s*:\\s*(\\d+)", options: .regularExpression) else { return 0 }
        let match = String(self[range])
        guard let colonRange = match.range(of: ": ") else { return 0 }
        let numStr = match[colonRange.upperBound...].trimmingCharacters(in: .whitespaces)
        return Int64(numStr) ?? 0
    }

    func extractDouble(_ key: String) -> Double {
        guard let range = self.range(of: "\"\(key)\"\\s*:\\s*([\\d.eE+-]+)", options: .regularExpression) else { return 0 }
        let match = String(self[range])
        guard let colonRange = match.range(of: ": ") else { return 0 }
        let numStr = match[colonRange.upperBound...].trimmingCharacters(in: .whitespaces)
        return Double(numStr) ?? 0
    }
}

File: android/src/main/kotlin/noise/protocol/Benchmark.kt
package noise.protocol

/**
 * Holds the result of a single benchmark measurement.
 *
 * @property name Human-readable name of the benchmark.
 * @property iterations Number of iterations executed during the measurement phase.
 * @property totalNs Total elapsed wall-clock time in nanoseconds.
 * @property opsPerSec Throughput expressed as operations per second.
 * @property avgNs Average time per operation in nanoseconds.
 */
data class BenchmarkResult(
    val name: String,
    val iterations: Int,
    val totalNs: Long,
    val opsPerSec: Double,
    val avgNs: Double
)

/**
 * Compares a single benchmark between a baseline and the current run.
 *
 * @property name The benchmark name.
 * @property baselineOpsPerSec Baseline throughput (ops/sec).
 * @property currentOpsPerSec Current throughput (ops/sec).
 * @property speedup Ratio of current to baseline throughput (`> 1.0` means faster).
 * @property improved `true` if [speedup] is greater than `1.0`.
 */
data class BenchmarkComparison(
    val name: String,
    val baselineOpsPerSec: Double,
    val currentOpsPerSec: Double,
    val speedup: Double,
    val improved: Boolean
)

/**
 * A report comparing a set of benchmark results against a baseline.
 *
 * @property comparisons The individual benchmark comparisons.
 */
data class ComparisonReport(
    val comparisons: List<BenchmarkComparison>
) {
    /**
     * Formats the comparison data as a human-readable ASCII table with columns
     * for benchmark name, baseline ops/sec, current ops/sec, speedup ratio,
     * and a status indicator (FASTER / SLOWER / SAME).
     *
     * @return The formatted summary string.
     */
    fun toSummary(): String {
        val sb = StringBuilder()
        sb.appendLine("%-30s %15s %15s %10s %8s".format("Benchmark", "Baseline", "Current", "Speedup", "Status"))
        sb.appendLine("-".repeat(83))
        for (c in comparisons) {
            val status = when {
                c.speedup > 1.05 -> "✓ FASTER"
                c.speedup < 0.95 -> "✗ SLOWER"
                else -> "= SAME"
            }
            sb.appendLine("%-30s %13.1f %13.1f %9.2fx %8s".format(
                c.name, c.baselineOpsPerSec, c.currentOpsPerSec, c.speedup, status
            ))
        }
        return sb.toString()
    }
}

/**
 * A collection of benchmark results for a single platform and point in time.
 *
 * Supports JSON serialization ([toJson] / [fromJson]) and comparison against
 * a baseline suite via [compareToBaseline].
 *
 * @property platform The platform identifier (e.g. `"JVM 17"`, `"Android API 34"`).
 * @property timestamp ISO 8601 timestamp of when the suite was run.
 * @property results The individual [BenchmarkResult] entries.
 * @see BenchmarkRunner
 * @see ComparisonReport
 */
data class BenchmarkSuite(
    val platform: String,
    val timestamp: String,
    val results: List<BenchmarkResult>
) {
    /**
     * Serializes this suite to a JSON string.
     *
     * @return The JSON representation.
     */
    fun toJson(): String {
        val sb = StringBuilder()
        sb.appendLine("{")
        sb.appendLine("""  "platform": "${platform.jsonEscape()}",""")
        sb.appendLine("""  "timestamp": "${timestamp.jsonEscape()}",""")
        sb.appendLine("""  "results": [""")
        results.forEachIndexed { i, r ->
            sb.appendLine("    {")
            sb.appendLine("""      "name": "${r.name.jsonEscape()}",""")
            sb.appendLine("""      "iterations": ${r.iterations},""")
            sb.appendLine("""      "totalNs": ${r.totalNs},""")
            sb.appendLine("""      "opsPerSec": ${r.opsPerSec},""")
            sb.appendLine("""      "avgNs": ${r.avgNs}""")
            sb.append("    }")
            if (i < results.size - 1) sb.appendLine(",") else sb.appendLine()
        }
        sb.appendLine("  ]")
        sb.append("}")
        return sb.toString()
    }

    /**
     * Writes the JSON serialization of this suite to a [file].
     *
     * @param file The destination file.
     */
    fun saveToFile(file: java.io.File) {
        file.writeText(toJson())
    }

    /**
     * Compares each result in this suite against the matching result (by name)
     * in the [baseline] suite and produces a [ComparisonReport].
     *
     * If a benchmark name exists in this suite but not in the [baseline],
     * the current ops/sec is used as the baseline (speedup = 1.0).
     *
     * @param baseline The baseline suite to compare against.
     * @return A [ComparisonReport] with per-benchmark speedup data.
     */
    fun compareToBaseline(baseline: BenchmarkSuite): ComparisonReport {
        val baselineByName = baseline.results.associateBy { it.name }
        val comparisons = results.map { current ->
            val base = baselineByName[current.name]
            val baseOps = base?.opsPerSec ?: current.opsPerSec
            val speedup = if (baseOps > 0) current.opsPerSec / baseOps else 1.0
            BenchmarkComparison(
                name = current.name,
                baselineOpsPerSec = baseOps,
                currentOpsPerSec = current.opsPerSec,
                speedup = speedup,
                improved = speedup > 1.0
            )
        }
        return ComparisonReport(comparisons)
    }

    companion object {
        /**
         * Deserializes a [BenchmarkSuite] from a JSON string.
         *
         * @param json The JSON string (as produced by [toJson]).
         * @return The deserialized [BenchmarkSuite].
         */
        fun fromJson(json: String): BenchmarkSuite {
            val platform = json.extractString("platform")
            val timestamp = json.extractString("timestamp")
            val results = mutableListOf<BenchmarkResult>()

            val resultsArray = json.substringAfter("\"results\":")
                .substringAfter("[").substringBeforeLast("]")
            val objects = resultsArray.split(Regex("\\}\\s*,\\s*\\{"))

            for (obj in objects) {
                val cleaned = obj.trim().removePrefix("{").removeSuffix("}")
                if (cleaned.isBlank()) continue
                results.add(
                    BenchmarkResult(
                        name = cleaned.extractString("name"),
                        iterations = cleaned.extractInt("iterations"),
                        totalNs = cleaned.extractLong("totalNs"),
                        opsPerSec = cleaned.extractDouble("opsPerSec"),
                        avgNs = cleaned.extractDouble("avgNs")
                    )
                )
            }

            return BenchmarkSuite(platform, timestamp, results)
        }

        /**
         * Loads a [BenchmarkSuite] from a JSON [file].
         *
         * @param file The source file containing JSON.
         * @return The deserialized [BenchmarkSuite].
         */
        fun loadFromFile(file: java.io.File): BenchmarkSuite = fromJson(file.readText())
    }
}

private fun String.jsonEscape(): String = replace("\\", "\\\\").replace("\"", "\\\"")

private fun String.extractString(key: String): String =
    Regex(""""$key"\s*:\s*"([^"]*?)"""").find(this)?.groupValues?.get(1) ?: ""

private fun String.extractInt(key: String): Int =
    Regex(""""$key"\s*:\s*(\d+)""").find(this)?.groupValues?.get(1)?.toInt() ?: 0

private fun String.extractLong(key: String): Long =
    Regex(""""$key"\s*:\s*(\d+)""").find(this)?.groupValues?.get(1)?.toLong() ?: 0L

private fun String.extractDouble(key: String): Double =
    Regex(""""$key"\s*:\s*([\d.E+-]+)""").find(this)?.groupValues?.get(1)?.toDouble() ?: 0.0

/**
 * Runs micro-benchmarks with a configurable warmup and measurement phase.
 *
 * The warmup phase allows the JIT compiler to optimize hotspots before
 * measurement begins.
 *
 * @param warmupIterations Number of iterations to run before measuring (default: 100).
 * @param measureIterations Number of iterations to measure (default: 1000).
 * @see BenchmarkResult
 * @see BenchmarkSuite
 */
class BenchmarkRunner(
    private val warmupIterations: Int = 100,
    private val measureIterations: Int = 1000
) {
    /**
     * Executes a benchmark: runs the warmup phase, then measures [block]
     * for [measureIterations] iterations.
     *
     * @param name A descriptive name for this benchmark.
     * @param block The code to benchmark (called once per iteration).
     * @return A [BenchmarkResult] containing timing statistics.
     */
    fun run(name: String, block: () -> Unit): BenchmarkResult {
        // Warmup phase — let JIT compile hotspots
        repeat(warmupIterations) { block() }

        // Measure phase
        val start = System.nanoTime()
        repeat(measureIterations) { block() }
        val totalNs = System.nanoTime() - start

        val avgNs = totalNs.toDouble() / measureIterations
        val opsPerSec = if (totalNs > 0) measureIterations.toDouble() / (totalNs / 1_000_000_000.0) else 0.0

        return BenchmarkResult(
            name = name,
            iterations = measureIterations,
            totalNs = totalNs,
            opsPerSec = opsPerSec,
            avgNs = avgNs
        )
    }
}
